src/engine/vni.rs

```diff
@@ -70,7 +70,7 @@ impl Vni {
         let mut is_first_match = true;
         for i in 0..buffer_len {
             let ch = self.buffer[i];
-            let next_ch = if i == buffer_len - 1 {
+            let next_ch = if i + 1 == buffer_len {
```

Comment:
I don't think this is necessary because the loop is from 0 to `buffer_len` so if `buffer_len` is 0 then it won't loop at all:) But I'll accept this to keep consistency with your other fixes.